### PR TITLE
Add Mcore microbatches calculator supports

### PR DESCRIPTION
**Pull request by BoxiangW**
# What does this PR do ?

Add Mcore microbatches calculator supports. Added backward compatibility. Merged `init_num_mb_calculator` and `reconfigure_num_mb_calculator`

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by BoxiangW** [comment]
This PR is an replacement of https://github.com/NVIDIA/NeMo/pull/9838 due to signoff issue.

**Comment by BoxiangW** [comment]
Adding decreasing related support in separate PR later

File: nemo/lightning/pytorch/plugins/data_sampler.py [hunk 1]
```diff
@@ -3,6 +3,16 @@
...
 
+try:
+    from megatron.core.num_microbatches_calculator import (
+        get_current_global_batch_size,
+        get_num_microbatches,
+        update_num_microbatches,
+    )
+
+except (ImportError, ModuleNotFoundError):
+    from apex.transformer.pipeline_parallel.utils import get_current_global_batch_size, get_num_microbatches
```
Comment (ashors1) [inline nemo/lightning/pytorch/plugins/data_sampler.py:14]: Can we move the imports back to where they were? We do the imports in-line to support configuring things locally without requiring all dependencies needed to run

**Comment by ashors1** [comment]
Besides the comment about the in-line imports, LGTM

File: nemo/lightning/pytorch/plugins/data_sampler.py (hunk 1 above)
Comment (BoxiangW) [reply nemo/lightning/pytorch/plugins/data_sampler.py]: Could you take a look again? I have made the changes.

**Review by terrykong — CHANGES_REQUESTED** [review]
I haven't reviewed each instance, but it seems like there are two categories of imports:
1. Where it's okay if the import doesn't succeed
```py
try:
    from megatron.core.num_microbatches_calculator import get_num_microbatches

except (ImportError, ModuleNotFoundError):
    try:
        from apex.transformer.pipeline_parallel.utils import get_num_microbatches

        HAVE_APEX = True
    except (ImportError, ModuleNotFoundError):

        HAVE_APEX = False
```
2. It's not okay if the import doesn't succeed
```py
try:
    from megatron.core.num_microbatches_calculator import update_num_microbatches

except (ImportError, ModuleNotFoundError):
    from apex.transformer.pipeline_parallel.utils import update_num_microbatches
```

But it seems like some modules like [megatron_nsfw_clip_models.py](https://github.com/NVIDIA/NeMo/pull/9968/files#diff-99c51b2778549b89eddb3a29e3a25ed8a430a3af0b4d1b9d554ff74b5c730b31) will fail at import time, but others like [controlnet.py](https://github.com/NVIDIA/NeMo/pull/9968/files#diff-aa45adaf14536aa483a53a21556b40770df844213bce2b976a0ae6c53ac8d834) won't have any issue during import, but you will have an issue when using the classes in that module.

What do you think about standardizing this? It seems strange to error when importing on some modules, especially when the utilities in both are similar, e.g., both modules define models.

**Comment by BoxiangW** [comment]
@terrykong This is actually an legacy setting inherited from previous version. They use try except to avoid import error and use 
```
if not HAVE_MEGATRON_CORE:
            logging.warning(
                "megatron-core was not found. Please see the NeMo README for installation instructions: https://github.com/NVIDIA/NeMo#megatron-gpt."
            )
```
to log warnings. I am thinking try to avoid this part and log a version warning for not able to use mcore mb calculator like:
```
try:
    from megatron.core.num_microbatches_calculator import get_num_microbatches

except (ImportError, ModuleNotFoundError):
    logging.warning("Megatron num_microbatches_calculator not found, using Apex version.")
    from apex.transformer.pipeline_parallel.utils import get_num_microbatches
```

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 2]
```diff
@@ -50,6 +48,25 @@
...
+except (ImportError, ModuleNotFoundError):
+    logging.warning("Megatron num_microbatches_calculator not found, using Apex version.")
+    from apex.transformer.pipeline_parallel.utils import (
+        _reconfigure_microbatch_calculator as reconfigure_num_microbatches_calculator,
+    )
+    from apex.transformer.pipeline_parallel.utils import (
+        get_current_global_batch_size,
+        get_micro_batch_size,
+        get_num_microbatches,
+    )
```
Comment (terrykong) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:68]: You should be able to combine these
```suggestion
    logging.warning("Megatron num_microbatches_calculator not found, using Apex version.")
    from apex.transformer.pipeline_parallel.utils import (
        get_current_global_batch_size,
        get_micro_batch_size,
        get_num_microbatches,
        _reconfigure_microbatch_calculator as reconfigure_num_microbatches_calculator,
    )
```

File: nemo/collections/nlp/models/language_modeling/megatron_t5_sft_model.py [hunk 3]
```diff
@@ -43,6 +41,25 @@
...
+
+except (ImportError, ModuleNotFoundError):
+    logging.warning("Megatron num_microbatches_calculator not found, using Apex version.")
+    from apex.transformer.pipeline_parallel.utils import (
+        _reconfigure_microbatch_calculator as reconfigure_num_microbatches_calculator,
+    )
+    from apex.transformer.pipeline_parallel.utils import (
+        get_current_global_batch_size,
+        get_micro_batch_size,
+        get_num_microbatches,
```
Comment (terrykong) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_sft_model.py:60]: same comment about combining these

**Review by terrykong — APPROVED** [review]

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py (hunk 2 above)
Comment (BoxiangW) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:68]: This is the styled by isort on my end. Seems like isort is defaulting this style

Comment (terrykong) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:68]: interesting. please disregard my review :)

**Pull request merged**